src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/AesCcm.Unix.cs

```diff
@@ -0,0 +1,211 @@
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpCipherSetKeyAndIV(ctx, ref MemoryMarshal.GetReference(_keyHandle.Key), ref MemoryMarshal.GetReference(nonce), 1 /* encrypting */))
+                {
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                }
+
+                if (associatedData.Length != 0)
+                {
+                    // pass length of the plaintext
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref nullRef, out _, ref nullRef, plaintext.Length))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref nullRef, out _, ref MemoryMarshal.GetReference(associatedData), associatedData.Length))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+                }
+
+                int ciphertextBytesWritten;
+                if (plaintext.Length != 0)
+                {
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref MemoryMarshal.GetReference(ciphertext), out ciphertextBytesWritten, ref MemoryMarshal.GetReference(plaintext), plaintext.Length))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+
+                    if (!Interop.Crypto.EvpCipherFinalEx(ctx, ref MemoryMarshal.GetReference(ciphertext.Slice(ciphertextBytesWritten)), out int bytesWritten))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+
+                    ciphertextBytesWritten += bytesWritten;
+                }
+                else
+                {
+                    byte nonEmptyArray = 123;
+                    if (!Interop.Crypto.EvpCipherUpdate(ctx, ref nonEmptyArray, out ciphertextBytesWritten, ref nonEmptyArray, plaintext.Length))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+
+                    if (!Interop.Crypto.EvpCipherFinalEx(ctx, ref nonEmptyArray, out int bytesWritten))
+                    {
+                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                    }
+
+                    // Should be 0 but let's still validate it
+                    ciphertextBytesWritten += bytesWritten;
+                }
+
+                if (ciphertextBytesWritten != plaintext.Length)
+                {
+                    // this shouldn't happen
+                    throw new CryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpAesCcmGetTag(ctx, ref MemoryMarshal.GetReference(tag), tag.Length))
```

Comment:
I'd also make it do the return check and throw, since that's just boilerplate.

Reply:
> I'd also make it do the return check and throw, since that's just boilerplate.

Building on this point:

When we first started doing Core on non-Windows we started with doing direct P/invokes, then we found that we sometimes needed a shim, then we moved to always using the shim.  After that, we started deciding that we had no real reason to make the code in classes like this look directly like the native thing.  So

```C#
[DllImport(etc)]
internal static extern bool EvpAesCcmGetTag(SafeEvpCipherCtxHandle ctx, ref byte tag, int tagLen);
```

started to become

```C#
[DllImport(etc)]
private static extern bool CryptoNative_EvpAesCcmGetTag(SafeEvpCipherCtxHandle ctx, ref byte tag, int tagLen);

internal static void EvpAesCcmGetTag(SafeEvpCipherCtxHandle ctx, Span<byte> tag)
{
    if (!CryptoNative_EvpAesCcmGetTag(ctx, ref MemoryMarshal.GetReference(tag), tag.Length))
    {
        throw CreateOpenSslCryptographicException();
    }
}
```

Now a) the need to throw is always checked, b) the exception is always consistent, c) the pointer and length are less likely to get torn apart and d) the caller code looks more like C# (and therefore has less "if failed fail" boilerplate on every line)

We didn't go proactively clean up old things once we started doing that, so there's still a lot of "this looks like C" code floating around.

For anything that is just going to do the "throw an exception based on the error code" path you should make a helper in the Interop.Crypto class and make the p/invoke private.  Similar for things which need the ref and length.  If you have cases where you are going to do something different in the failure case(s) then that's either directly to the p/invoke or just a different helper.